### PR TITLE
Remove intermediate SearchLookup classes

### PR DESCRIPTION
**Pull request by romseygeek**
SearchLookup has two intermediate classes, DocMap and StoredFieldsLookup, that
are simple factories for their Leaf implementations.  They are never accessed 
outside SearchLookup, with the exception of two calls on DocMap that can be
easily refactored.  This commit removes them, making SearchLookup.getLeafSearchLookup
directly responsible for creating the leaf lookups.

### CODE REVIEW

**Comment by romseygeek** [comment]
This is a small tidy up that makes it easier to work out what is accessed where in SearchLookup.  It is a precursor to another PR that moves field reference tracking out of SearchLookup directly and into the fieldtype lookup function, so that we can track references on calculated fields.

File: modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java [hunk 1]
```diff
@@ -453,13 +453,13 @@ private static DoubleValuesSource getDocValueSource(String variable, SearchLooku
...
         String fieldname = parts[1].text;
-        MappedFieldType fieldType = lookup.doc().fieldType(fieldname);
+        MappedFieldType fieldType = lookup.fieldType(fieldname);
 
         if (fieldType == null) {
             throw new ParseException("Field [" + fieldname + "] does not exist in mappings", 5);
         }
 
-        IndexFieldData<?> fieldData = lookup.doc().getForField(fieldType);
+        IndexFieldData<?> fieldData = lookup.getForField(fieldType);
```
Comment (javanna) [inline modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java:462]: it is odd that we need two additional methods in SearchLookup. Here we could have a single getForField(String) but I guess in other places that would not work?

File: server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java [hunk 2]
```diff
@@ -37,14 +37,18 @@
 @SuppressWarnings({"unchecked", "rawtypes"})
 public class LeafStoredFieldsLookup implements Map<Object, Object> {
 
+    public interface FieldsReader {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java:40]: I was not able to follow why these changes are needed. Wouldn't everything keep working if context.reader() is provided as a constructor argument as before?

**Review by javanna — COMMENTED** [review]
left two small comments

File: server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java:40]: This one is a pain - it's because some tests pass `null` as a context, and so we can't necessarily call `context.reader()`.  I initially changed it to just pass LeafReaderContext instead and then read by calling `context.reader().document()` but that falls foul of a test that was using mocking to build a fake Reader.  You can't mock LeafReaderContext because it's final.  Hence this refactoring - it does mean that we can remove the mocking though, which makes the test a lot easier to read.

File: modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java (hunk 1 above)
Comment (romseygeek) [reply modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java:462]: We may be able to remove them entirely, let me try!

Comment (romseygeek) [reply modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java:462]: Alas, we need both here - we need access to both the field data and the MappedFieldType

File: server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java (hunk 2 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java:40]: Curious what tests were failing because of a `null` context? To me it looks like the logic was just moved around, so I'm surprised we'd have a different in behavior (I think I'm missing something though...)

**Review by jtibshirani — COMMENTED** [review]
This looks good to me, just left one question.

File: server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java:40]: It's the change in `FlattenedFieldLookupTests` from `LeafDocLookup docLookup = searchLookup.doc().getLeafDocLookup(null);` to `LeafDocLookup docLookup = searchLookup.getLeafSearchLookup(null).doc();`

Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java:40]: It seems to me that LeafReaderContext should never be null, and this specific test provides it null because it plugs in its own field data impl which makes the leaf reader context irrelevant. Then it sounds like we could simply find a way for this test to provide any non null value, rather than allowing for null in this way?

Would providing `Lucene.emptyReader(0).getContext()` work?

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java:40]: Luca and I spoke offline about this, and agreed that the removal of the complicated mocking in FlattenedFieldLookupTests makes this change worth doing.

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java:40]: Do you mean that we'll switch to creating a real context in `FlattenedFieldLookupTests` or keep this change? I don't have a strong opinion, and will approve so merging doesn't block on me. Last note: if we do end up sticking with the current strategy, maybe we could use `CheckedBiConsumer` to avoid introducing a new interface (with a very general name).

**Review by jtibshirani — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java]: We'll keep this change, and I have replaced the new interface with a CheckedBiConsumer.  Thanks for the reviews!
